Add OMG WASM tracing debug code

justinmichaud opened the pull request:
#### 37b14a877442d6b80bac19213962e5376267a110
<pre>
Add OMG WASM tracing debug code
<a href="https://bugs.webkit.org/show_bug.cgi?id=254711">https://bugs.webkit.org/show_bug.cgi?id=254711</a>
rdar://107394100

Reviewed by Yusuke Suzuki.

Add some debugging tools to the OMG generator to trace WASM execution
at runtime.

* Source/JavaScriptCore/b3/B3Value.cpp:
(JSC::B3::Value::generateCompilerConstructionSite):
* Source/JavaScriptCore/b3/B3Value.h:
* Source/JavaScriptCore/wasm/WasmAirIRGeneratorBase.h:
(JSC::Wasm::AirIRGeneratorBase::didPopValueFromStack):
* Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp:
(JSC::Wasm::B3IRGenerator::didPopValueFromStack):
(JSC::Wasm::B3IRGenerator::makePushVariable):
(JSC::Wasm::B3IRGenerator::push):
(JSC::Wasm::B3IRGenerator::addRefFunc):
(JSC::Wasm::B3IRGenerator::getLocal):
(JSC::Wasm::B3IRGenerator::traceValue):
(JSC::Wasm::B3IRGenerator::traceCF):
(JSC::Wasm::B3IRGenerator::setLocal):
(JSC::Wasm::B3IRGenerator::getGlobal):
(JSC::Wasm::B3IRGenerator::setGlobal):
(JSC::Wasm::B3IRGenerator::addLoop):
(JSC::Wasm::B3IRGenerator::addIf):
(JSC::Wasm::B3IRGenerator::addElseToUnreachable):
(JSC::Wasm::B3IRGenerator::addTry):
(JSC::Wasm::B3IRGenerator::emitCatchImpl):
(JSC::Wasm::B3IRGenerator::addThrow):
(JSC::Wasm::B3IRGenerator::addReturn):
(JSC::Wasm::B3IRGenerator::addBranch):
(JSC::Wasm::B3IRGenerator::addEndToUnreachable):
(JSC::Wasm::B3IRGenerator::addCall):
(JSC::Wasm::B3IRGenerator::addCallIndirect):
(JSC::Wasm::B3IRGenerator::addCallRef):
* Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
(JSC::Wasm::BBQJIT::didPopValueFromStack):
* Source/JavaScriptCore/wasm/WasmFunctionParser.h:
(JSC::Wasm::FunctionParser&lt;Context&gt;::parseExpression):
* Source/JavaScriptCore/wasm/WasmLLIntGenerator.cpp:
(JSC::Wasm::LLIntGenerator::didPopValueFromStack):
(JSC::Wasm::LLIntGenerator::atomicStore):
* Source/JavaScriptCore/wasm/WasmOperations.cpp:
(JSC::Wasm::loadValuesIntoBuffer):
* Source/JavaScriptCore/wasm/WasmOperationsInlines.h:
* Source/JavaScriptCore/wasm/generateWasmB3IRGeneratorInlinesHeader.py:
(CodeGenerator.makeResult):

Canonical link: <a href="https://commits.webkit.org/262784@main">https://commits.webkit.org/262784@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e2236fd5b22996b62c058cec9a3cbda24dd62dd9

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/2444 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/2443 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/2566 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/3788 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/2483 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/2583 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/2531 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/2166 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/2466 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/2248 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/2197 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/3573 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/213 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/2182 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/2043 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/2041 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/2307 "Built successfully") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/2204 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/3364 "260 api tests failed or timed out") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/2340 "Built successfully and passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/2232 "Passed tests") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/2024 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/2519 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/2227 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/2185 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/567 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/639 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/2199 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/2571 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/2352 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/679 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/wasm/WasmFunctionParser.h:
```diff
@@ -2434,13 +2437,15 @@ FOR_EACH_WASM_MEMORY_STORE_OP(CREATE_CASE)
+            m_context.didPopValueFromStack(arg, "Call"_s);
         }
         m_expressionStack.shrink(firstArgumentIndex);
+        args.reverse();
```
Why not storing the right index of `args` directly instead of `uncheckedAppend`? We already know the size, so we can just `resize` and store it instead of calling `reverse` here.

Constellation commented on Source/JavaScriptCore/wasm/WasmFunctionParser.h:
```diff
@@ -2085,14 +2087,15 @@ FOR_EACH_WASM_MEMORY_STORE_OP(CREATE_CASE)
+                m_context.didPopValueFromStack(arg, "StructNew*"_s);
             }
             m_expressionStack.shrink(firstArgumentIndex);
+            args.reverse();
```
Ditto.

Constellation commented on Source/JavaScriptCore/wasm/WasmFunctionParser.h:
```diff
@@ -1879,13 +1880,14 @@ FOR_EACH_WASM_MEMORY_STORE_OP(CREATE_CASE)
+                m_context.didPopValueFromStack(arg, "GC ArrayNew"_s);
             }
             m_expressionStack.shrink(firstArgumentIndex);
+            args.reverse();
```
Ditto.

Constellation commented on Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp:
```diff
@@ -787,21 +798,24 @@ class B3IRGenerator {
-    }
-
-    ExpressionType push(B3::Type type)
+    ExpressionType makePushVariable(B3::Type type)
```
I wonder if we should just name it `getPushVariable`, or `getStackVariable` since the purpose of this function is not making it: if it already exists, we just return it without "making" it. So calling it `make` seems wrong to me.

Constellation reviewed (approved):
r=me with comments.

justinmichaud commented on Source/JavaScriptCore/wasm/WasmFunctionParser.h:
```diff
@@ -2563,14 +2566,15 @@ FOR_EACH_WASM_MEMORY_STORE_OP(CREATE_CASE)
         WASM_PARSER_FAIL_IF(argumentCount > m_expressionStack.size(), "call_ref expects ", argumentCount, " arguments, but the expression stack currently holds ", m_expressionStack.size(), " values");
 
         Vector<ExpressionType> args;
-        WASM_PARSER_FAIL_IF(!args.tryReserveCapacity(argumentCount + 1), "can't allocate enough memory for ", argumentCount, " call_indirect arguments");
```
This looks like it is not needed, right?

Constellation replied:
We need to check this first to allocate capacity with fail-safe way. And then we should call resize to make Vector size matched (not capacity).

justinmichaud replied:
I meant the +1

Constellation replied:
Yeah, looks like we do not need it.

Constellation commented on Source/JavaScriptCore/wasm/WasmFunctionParser.h:
```diff
@@ -1879,11 +1880,11 @@ FOR_EACH_WASM_MEMORY_STORE_OP(CREATE_CASE)
             WASM_PARSER_FAIL_IF(!args.tryReserveCapacity(argc), "can't allocate enough memory for array.new_fixed ", argc, " values");
```
This is just reserving a capacity. So args.size() is still 0. We need to call `args.resize(...)` to make the size matching against the arguments size.

Constellation replied:
Also, since `args` is definitely empty before this, we can use `tryReserveInitialCapacity` instead.

Constellation commented on Source/JavaScriptCore/wasm/WasmFunctionParser.h:
```diff
@@ -2085,12 +2086,12 @@ FOR_EACH_WASM_MEMORY_STORE_OP(CREATE_CASE)
             size_t firstArgumentIndex = m_expressionStack.size() - structType->fieldCount();
             WASM_PARSER_FAIL_IF(!args.tryReserveCapacity(structType->fieldCount()), "can't allocate enough memory for struct.new ", structType->fieldCount(), " values");
```
Ditto.

Constellation commented on Source/JavaScriptCore/wasm/WasmFunctionParser.h:
```diff
@@ -2085,12 +2086,12 @@ FOR_EACH_WASM_MEMORY_STORE_OP(CREATE_CASE)
                 WASM_VALIDATOR_FAIL_IF(!isSubtype(arg.type(), fieldType), "argument type mismatch in struct.new, got ", arg.type(), ", expected ", fieldType);
-                args.uncheckedAppend(arg);
-                m_context.didPopValueFromStack();
+                args[args.size() - i - 1] = arg;
```
We need to call `resize` before this.

Constellation commented on Source/JavaScriptCore/wasm/WasmFunctionParser.h:
```diff
@@ -1879,11 +1880,11 @@ FOR_EACH_WASM_MEMORY_STORE_OP(CREATE_CASE)
                 WASM_VALIDATOR_FAIL_IF(!isSubtype(arg.type(), elementType), "argument type mismatch in array.new_fixed, got ", arg.type(), ", expected a subtype of ", elementType);
-                args.uncheckedAppend(arg);
-                m_context.didPopValueFromStack();
+                args[args.size() - i - 1] = arg;
```
We need to call resize before this.

Constellation commented on Source/JavaScriptCore/wasm/WasmFunctionParser.h:
```diff
@@ -2434,11 +2435,12 @@ FOR_EACH_WASM_MEMORY_STORE_OP(CREATE_CASE)
         size_t firstArgumentIndex = m_expressionStack.size() - calleeSignature.argumentCount();
         Vector<ExpressionType> args;
         WASM_PARSER_FAIL_IF(!args.tryReserveCapacity(calleeSignature.argumentCount()), "can't allocate enough memory for call's ", calleeSignature.argumentCount(), " arguments");
```
Ditto.

Constellation commented on Source/JavaScriptCore/wasm/WasmFunctionParser.h:
```diff
@@ -2434,11 +2435,12 @@ FOR_EACH_WASM_MEMORY_STORE_OP(CREATE_CASE)
+            size_t stackIndex = m_expressionStack.size() - i - 1;
+            TypedExpression arg = m_expressionStack.at(stackIndex);
+            WASM_VALIDATOR_FAIL_IF(!isSubtype(arg.type(), calleeSignature.argumentType(calleeSignature.argumentCount() - i - 1)), "argument type mismatch in call, got ", arg.type(), ", expected ", calleeSignature.argumentType(calleeSignature.argumentCount() - i - 1));
+            args[args.size() - i - 1] = arg;
```
Ditto.

Constellation commented on Source/JavaScriptCore/wasm/WasmFunctionParser.h:
```diff
@@ -2502,12 +2504,13 @@ FOR_EACH_WASM_MEMORY_STORE_OP(CREATE_CASE)
         Vector<ExpressionType> args;
         WASM_PARSER_FAIL_IF(!args.tryReserveCapacity(argumentCount), "can't allocate enough memory for ", argumentCount, " call_indirect arguments");
```
Ditto.

Constellation commented on Source/JavaScriptCore/wasm/WasmFunctionParser.h:
```diff
@@ -2502,12 +2504,13 @@ FOR_EACH_WASM_MEMORY_STORE_OP(CREATE_CASE)
+            TypedExpression arg = m_expressionStack.at(stackIndex);
+            if (i > 0)
+                WASM_VALIDATOR_FAIL_IF(!isSubtype(arg.type(), calleeSignature.argumentType(argumentCount - i - 1)), "argument type mismatch in call_indirect, got ", arg.type(), ", expected ", calleeSignature.argumentType(argumentCount - i - 1));
+            args[args.size() - i - 1] = arg;
```
Ditto.

Constellation commented on Source/JavaScriptCore/wasm/WasmFunctionParser.h:
```diff
@@ -2777,10 +2781,10 @@ FOR_EACH_WASM_MEMORY_STORE_OP(CREATE_CASE)
         Vector<ExpressionType> args;
         WASM_PARSER_FAIL_IF(!args.tryReserveCapacity(exceptionSignature.argumentCount()), "can't allocate enough memory for throw's ", exceptionSignature.argumentCount(), " arguments");
```
Ditto.

Constellation commented on Source/JavaScriptCore/wasm/WasmFunctionParser.h:
```diff
@@ -2777,10 +2781,10 @@ FOR_EACH_WASM_MEMORY_STORE_OP(CREATE_CASE)
-            m_context.didPopValueFromStack();
+            TypedExpression arg = m_expressionStack.at(m_expressionStack.size() - i - 1);
+            WASM_VALIDATOR_FAIL_IF(arg.type() != exceptionSignature.argumentType(exceptionSignature.argumentCount() - i - 1), "The exception being thrown expects the argument at index ", i, " to be ", exceptionSignature.argumentType(exceptionSignature.argumentCount() - i - 1), " but argument has type ", arg.type());
+            args[args.size() - i - 1] = arg;
```
Ditto

Commit:
Add OMG WASM tracing debug code
https://bugs.webkit.org/show_bug.cgi?id=254711
rdar://107394100

Reviewed by Yusuke Suzuki.

Add some debugging tools to the OMG generator to trace WASM execution
at runtime.

* Source/JavaScriptCore/b3/B3Value.cpp:
(JSC::B3::Value::generateCompilerConstructionSite):
* Source/JavaScriptCore/b3/B3Value.h:
* Source/JavaScriptCore/wasm/WasmAirIRGeneratorBase.h:
(JSC::Wasm::AirIRGeneratorBase::didPopValueFromStack):
* Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp:
(JSC::Wasm::B3IRGenerator::didPopValueFromStack):
(JSC::Wasm::B3IRGenerator::makePushVariable):
(JSC::Wasm::B3IRGenerator::push):
(JSC::Wasm::B3IRGenerator::addRefFunc):
(JSC::Wasm::B3IRGenerator::getLocal):
(JSC::Wasm::B3IRGenerator::traceValue):
(JSC::Wasm::B3IRGenerator::traceCF):
(JSC::Wasm::B3IRGenerator::setLocal):
(JSC::Wasm::B3IRGenerator::getGlobal):
(JSC::Wasm::B3IRGenerator::setGlobal):
(JSC::Wasm::B3IRGenerator::addLoop):
(JSC::Wasm::B3IRGenerator::addIf):
(JSC::Wasm::B3IRGenerator::addElseToUnreachable):
(JSC::Wasm::B3IRGenerator::addTry):
(JSC::Wasm::B3IRGenerator::emitCatchImpl):
(JSC::Wasm::B3IRGenerator::addThrow):
(JSC::Wasm::B3IRGenerator::addReturn):
(JSC::Wasm::B3IRGenerator::addBranch):
(JSC::Wasm::B3IRGenerator::addEndToUnreachable):
(JSC::Wasm::B3IRGenerator::addCall):
(JSC::Wasm::B3IRGenerator::addCallIndirect):
(JSC::Wasm::B3IRGenerator::addCallRef):
* Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
(JSC::Wasm::BBQJIT::didPopValueFromStack):
* Source/JavaScriptCore/wasm/WasmFunctionParser.h:
(JSC::Wasm::FunctionParser<Context>::parseExpression):
* Source/JavaScriptCore/wasm/WasmLLIntGenerator.cpp:
(JSC::Wasm::LLIntGenerator::didPopValueFromStack):
(JSC::Wasm::LLIntGenerator::atomicStore):
* Source/JavaScriptCore/wasm/WasmOperations.cpp:
(JSC::Wasm::loadValuesIntoBuffer):
* Source/JavaScriptCore/wasm/WasmOperationsInlines.h:
* Source/JavaScriptCore/wasm/generateWasmB3IRGeneratorInlinesHeader.py:
(CodeGenerator.makeResult):

Canonical link: https://commits.webkit.org/262784@main
